[OMID-70] World time based timestamp generator

ohadshacham opened the pull request:
Adding world time base timestamp generator. The generator maintains monotonicity and persists the maximal timestamp in each interval.
    
The motivation for this commit is the integration with Apache Phoenix.
Phoenix's customers requires mapping from timestamp to world time in order to debug their logs.
    
This commit also contains augmentation of the protobuf in order to partially serialize transaction. This is also for Phoenix usage.

francisco-perez-sorrosal committed:
[OMID-64] Fix location of protobuf tgz in travis config

Location has changed to
https://github.com/google/protobuf/releases/download/v2.5.0/protobuf-2.5.0.tar.gz

Change-Id: Icd7740ff23bf17142fb47bdc530178b0292b3a05

Commit:
Adding world base timestamp generator. The generator maintains monotonicity and persists the maximal timestamp in each interval.

The motivation for this commit is the integration with Apache Phoenix.
Phoenix's customers requires mapping from timestamp to world time in order to debug their logs.

This commit also contains augmentation of the proto in order to partially serialize transaction. This is also for Phoenix usage.

Commit:
cosmetics

Commit:
Fix the timestampType field to string and add a getter for the enum

ebortnik commented on common/src/main/proto/TSOProto.proto:
```diff
@@ -64,3 +64,8 @@ message HandshakeResponse {
     optional Capabilities serverCapabilities = 2;
 }
+
+message Transaction {
```
Transaction --> TransactionEncoding

Suggest to add a method to Omid TM to return this instance. We should keep the Omid internals unexposed to TAL.

ebortnik commented on tso-server/src/main/java/org/apache/omid/tso/CurrentTimestampOracleImpl.java:
```diff
@@ -0,0 +1,171 @@
+/*
```
Suggest to rename: CurrentTimestampOracleImpl (wrong English) --> WorldClockOracleImpl

ebortnik commented on tso-server/src/main/java/org/apache/omid/tso/CurrentTimestampOracleImpl.java:
```diff
@@ -0,0 +1,171 @@
+        }
+    }
+
+    static final long MAX_TX_PER_MS = 1_000_000; // 1 million
```
Suggest to move all state variables to beginning of class

ebortnik commented on tso-server/src/main/java/org/apache/omid/tso/CurrentTimestampOracleImpl.java:
```diff
@@ -0,0 +1,171 @@
+    private static final Logger LOG = LoggerFactory.getLogger(CurrentTimestampOracleImpl.class);
+
+    @VisibleForTesting
+    static class InMemoryTimestampStorage implements TimestampStorage {
```
Who uses this class? If this is just test code, please move down.

ebortnik commented on tso-server/src/main/java/org/apache/omid/tso/TSOModule.java:
```diff
@@ -43,7 +46,13 @@ protected void configure() {
         bind(TSOStateManager.class).to(TSOStateManagerImpl.class).in(Singleton.class);
-        bind(TimestampOracle.class).to(TimestampOracleImpl.class).in(Singleton.class);
+
+        if (config.getTimestampTypeEnum() == TIMESTAMP_TYPE.WORLD_TIME) {
```
Better to use switch

Commit:
[OMID-71] Omid's transaction manager detects conflicts at the
 cell level. Meaning that two concurrent transactions write to different
 columns at the same row do not conflict. This semantics is not suitable for
 Apache Phoenix which requires conflict detection at the row level. This
 commit augments Omid with row level conflict analysis.

ebortnik commented on tso-server/src/main/java/org/apache/omid/tso/CurrentTimestampOracleImpl.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    @Override
+    public void initialize() throws IOException {
```
What/where is the code for the failover case?

ohadshacham replied:
This is the code for the failover. It allocates the next interval and waits in case the next interval starts in the future. This can occur when the previous TSO already allocated the current interval and we need to wait it will over. The reason we are waiting is to guarantee that the timestamp really representing the world time the transaction was created/committed.

ebortnik commented on tso-server/src/main/java/org/apache/omid/tso/CurrentTimestampOracleImpl.java:
```diff
@@ -0,0 +1,171 @@
+        executor.execute(allocateTimestampsBatchTask);
+
+        // Waiting for the current epoch to start. Occurs in case of fallback when the previous TSO allocated the current time frame.
+        while ((System.currentTimeMillis() * MAX_TX_PER_MS) < this.lastTimestamp);
```
Busy wait - dangerous !! Please add short sleeps within the loop. 
Would you interrupt the process if the condition does not ever hold? 
Finally, please improve the comments - when you apply this code? Failover scenario?

ebortnik commented on tso-server/src/main/java/org/apache/omid/tso/CurrentTimestampOracleImpl.java:
```diff
@@ -0,0 +1,171 @@
+
+        if (currentMsFirstTimestamp >= nextAllocationThreshold) {
+                nextAllocationThreshold = Long.MAX_VALUE; // guarantees that only one will enter this branch (this is a sequential code so it should work)
+                executor.execute(allocateTimestampsBatchTask);
```
As per our decision, please switch to periodic renewal from on-demand.

ebortnik commented on tso-server/src/test/java/org/apache/omid/tso/TestCurrentTimestampOracle.java:
```diff
@@ -0,0 +1,113 @@
+/*
```
Please change the class name to include WorldTime

Commit:
Merge branch 'master' into WorldTimeStamp

Commit:
Integrating ebortnik comments from pull request #123

Commit:
adding additional comments

Commit:
Merge branch 'master' into WorldTimeStamp
